### DISCUSSION TITLE
Conversion of QCQP to QP with MOSEK and optimizer object

### DISCUSSION THREAD

**Discussion by Melanie-G** [yalmip/YALMIP#1367] [Possible bug or missing functionality]
Hi, 

I have encountered a problem when using MOSEK in combination with YALMIP for solving a quadratically constrained quadratic program in MATLAB. When calling optimize, the code works as expected, but when using the optimizer object, the quadratic constraints are ignored and the problem is converted into a quadratic program instead of a QCQP or conic program. I have attached a test script with a quadratic constraint, calling both optimize and the optimizer object. The quadratic constraint is violated using the latter method.

I have traced the code and noticed that in '\YALMIP\extras\@optimizer\optimizer.m' lines 171-177 a + is added to the user solver setting to enforce the selected solver, which is not done in the optimize call. This leads to LP/QP Mosek solvers in '\YALMIP\extras\selectsolver.m' lines 296-302 not being pruned and remaining in first place in the solvers struct, above the conic solvers, and thus being selected as solver in l.513.

The conversion to a QP can be prevented by not specifying the solver or by adding a constraint of the type (norm(x,2) <= a).

```
clear

A = [1, 2;
    2, 1];
B = [0; 1];
P = [1, 0.5;
    0.5, 1];

Q = eye(2);
R = 1;

n_x = 2;
n_u = 1;
N = 5;

x = sdpvar(n_x*ones(1,N),ones(1,N));
u = sdpvar(n_u*ones(1,N-1),ones(1,N-1));

constraints = [];
objective = 0;

for k = 1:(N-1)
    constraints = [constraints, x{:,k+1} == A * x{:,k} + B * u{:,k} ]; 
    objective = objective +  x{:,k}' * Q * x{:,k} + u{:,k}' * R * u{:,k}; 
end  

% Quadratic terminal cost and constraints
objective = objective + x{:,N}' * P * x{:,N};
constraints = [constraints, x{:,N}' * P * x{:,N} <= 1];

x0 = [5; 10];
constraints = [constraints, x{:,1} == x0 ];

MPC_output = { [u{:}], [x{:}] };

options = sdpsettings('solver','mosek','verbose',2);


%% Version 1 - optimize
diag = optimize(constraints,objective,options);
eval_qc_optimize = value(x{:,N})' * P * value(x{:,N});

%% Version 2 - optimizer
controller = optimizer(constraints,objective,options,[],MPC_output);
[output, flag, errMsg] = controller();

x_opt = output{2};
eval_qc_optimizer = x_opt(:,N)' * P * x_opt(:,N);
```

**Comment by johanlofberg** [comment]
You probably have to code the quadratic constraints using conic constraints, i.e. norm(chol(P)*x)<=1. since YALMIP cannot be sure that the constraint will remain a quadratic (and thus require cones) when parameters are fixed

**Reply by Melanie-G** [reply]
Thanks for your reply! 
Shouldn't a more versatile solver be used in the case where YALMIP cannot be sure whether the constraints are quadratic or linear, to solve both problem classes correctly? Also, why does the behaviour differ between the optimize call and the optimizer object?

**Reply by johanlofberg** [reply]
It's...complicated
